### PR TITLE
WIP - building the watcher

### PR DESCRIPTION
**Pull request by sbernheim**
WIP - this PR is reviewing progress made building out the watcher loop.

It should not be merged.

### CODE REVIEW

**Comment by sbernheim** [comment]
@paulcarlton-ww

File: pkg/internal/kubectl/kubectl.go [hunk 1]
```diff
@@ -14,91 +14,126 @@ See the License for the specific language governing permissions and
...
+	logger logr.Logger
+	path   string
 }
 
-// NewKubectl returns a Kubectl object for creating and running kubectl sub-commands.
-func NewKubectl(l logr.Logger) (kctl *Kubectl, err error) {
-	kctl = &Kubectl{
-		logger: l,
+// NewKubectl returns a Kubectl object for creating and running kubectl sub-commands
+func NewKubectl(logger logr.Logger) (factory Kubectl, err error) {
```
Comment (richardcase) [inline pkg/internal/kubectl/kubectl.go:50]: Why is this needed? kubectl shouldn't be used from inside a controller. client-go should give you everything you need.

File: pkg/internal/layers/layers.go [hunk 2]
```diff
@@ -1,10 +1,13 @@
+//Package layers xxx
+//go:generate mockgen -destination=mockLayers.go -package=layers -source=layers.go . Layer
 package layers
 
 import (
 	"context"
 	"fmt"
 	"time"
 
+	helmrelease "github.com/fluxcd/helm-operator/pkg/client/clientset/versioned"
```
Comment (richardcase) [inline pkg/internal/layers/layers.go:10]: Really not sure of using the HR client directly (previous comment about this)

File: pkg/internal/layers/layers.go [hunk 3]
```diff
@@ -18,36 +21,24 @@ import (
...
 	StatusPruning()
-	StatusDeployed()
+	StatusDeployed(reason, message string)
 	IsHold() bool
 	SetHold()
 	GetStatus() string
 	GetName() string
 	GetNamespace() string
 	GetLogger() logr.Logger
 	GetK8sClient() *kubernetes.Clientset
```
Comment (richardcase) [inline pkg/internal/layers/layers.go:40]: Are functions like this needed on the interface? To me it feels like they shouldn't be there.

File: pkg/internal/kubectl/kubectl.go (hunk 1 above)
Comment (sbernheim) [reply pkg/internal/kubectl/kubectl.go:50]: @richardcase - In this case, client-go does not give us everything we need.  Most of the logic that goes into applying YAML to the cluster lives in the `kubectl` code and is not server-side or broken out into a module that other projects can pick up.  If you look into similar open-source projects, they generally shell out to `kubectl` rather than re-implement all that logic themselves, and there's no other way currently to guarantee full compatibility.

I was also specifically directed to use `kubectl` in this manner.

File: pkg/internal/apply/layerApplier.go [hunk 4]
```diff
@@ -0,0 +1,580 @@
...
+func isFailedPhase(phase *helmopv1.HelmReleasePhase) bool {
+	switch *phase {
+	case helmopv1.HelmReleasePhaseFailed:
+		return true
+	default:
+		return false
+	}
+}
+
+func (a *KubectlLayerApplier) watchAddons(hrClient *helmrelease.Clientset, hrs []*helmopv1.HelmRelease, layer layers.Layer) *addonsTracker {
```
[the remark this answers is not in the document]
Comment (sbernheim) [reply pkg/internal/apply/layerApplier.go]: @richardcase - I'd be happy to discuss it.  I looked into a couple of alternatives, but if you have more suggestions to offer I'm definitely open to learning more.

**Comment by sbernheim** [comment]
Closing as per our architecture discussion with @richardcase this morning.
